### DISCUSSION TITLE
Reading binary data

### DISCUSSION THREAD

**Discussion by js-motocrane** [Serial-Studio/Serial-Studio#171] [Q&A] [answered]
Hi, 

I have used Serial-Studio before for a simple widget with one piece of ASCII data being sent and it worked great.

Now I have a project where I will be sending a bunch of different data in binary format with no delimiters and I'm wondering how or if I can use Serial-Studio for this. 

The packet will be 40 bytes long and have '*' as a start and end character. 

The data in between the start/end has some different formats, such as uint8, uint16, and int16.

Is this possible? 

Thanks.

**Comment by alex-spataru** [comment]
Hi! First of all, happy new year! Best wishes for anyone reading this thread!

I think reading binary data is possible, however, you will need to customize the frame parser code:

1. Download the latest [continuous](https://github.com/Serial-Studio/Serial-Studio/releases/tag/continuous) build.
2. Create your project as usual with the Project Editor window.
3. Click on the "Customize frame parser" button:

<img width="1180" alt="213913612-dc82abf4-9152-442f-9827-64af516164c1" src="https://github.com/Serial-Studio/Serial-Studio/assets/4225542/f7306e39-50ea-4c15-b379-29f770d66c22">

4. The following window will appear, here, you can add custom JavaScript code to separate incoming data:

<img width="752" alt="213913696-75cb85cc-2898-4faf-85cd-e4fd3339a369" src="https://github.com/Serial-Studio/Serial-Studio/assets/4225542/e5fe117a-a48a-41c9-8e07-4103e7bcbe77">

The basic idea is that you need to modify the `parse` function so that it returns an array with the values stored inside the binary data that is contained in your frame. Your value "indexes" in the project file should be consistent with the array that you generate with the `parse` function. 

Please consider that the JavaScript code that you write should be compatible with the [Qt JS engine](https://wiki.qt.io/JavaScript), which might differ a bit from the "normal" JavaScript code you would write for a web application.

Another **important** observation: it is best to have different characters or strings for the start and end sequences, I have not really tested the frame parsing algorithm using the same start & end sequences.

**Reply by alex-spataru** [reply]
Another thing. I must do tests, it is possible that the string received by the "parse" function will be corrupted when receiving non-ASCII characters/bytes. I'll notify you as soon as I do test or modify the project in order to be able to process binary data.

**Comment by alex-spataru** [comment, marked as answer]
Just a heads-up: starting from version 3.0.0, you can now create dashboards using hexadecimal or binary data. For more details, check out the [Hexadecimal ADC Example](https://github.com/Serial-Studio/Serial-Studio/tree/master/examples/HexadecimalADC).
